### PR TITLE
Update Form component and incorporate navigation

### PR DESCRIPTION
**Pull request by marcja**
This is a draft. It's not done.

### CODE REVIEW

File: client/src/App.test.js [hunk 1]
```diff
@@ -22,7 +22,7 @@ const {
 } = require("./__fixtures__/catalog");
 
 describe("App", () => {
-  it("calls API, stores catalog in Redux, renders Form", async () => {
+  it.skip("calls API, stores catalog in Redux, renders Form", async () => {
```
Comment (kebrahim) [inline client/src/App.test.js:25]: is the plan to un-skip these in a followup PR?

File: client/src/Routes.test.js [hunk 2]
```diff
@@ -1,25 +1,25 @@
...
+import Routes from "./Routes.jsx";
 
 jest.mock("react-i18next", () => ({
   useTranslation: () => ({ t: (key) => key }),
 }));
 
 describe("Routes", () => {
-  it("routes / to QuestionForm", async () => {
-    const { getByText } = render(
+  it.skip("routes / to Form", async () => {
```
Comment (kebrahim) [inline client/src/Routes.test.js:14]: looks like you fixed this test. can it be unskipped?

File: client/src/components/renderWith.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
+import "@testing-library/jest-dom/extend-expect";
```
Comment (kebrahim) [inline client/src/components/renderWith.js:1]: should we indicate somehow that this file is to only be used for tests? i'm not sure if there's a best practice for doing so, in react

File: client/src/components/Form.jsx [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    // validate that for each questionKey in form, there is an equivalent
+    // questionKey in responses
+    history.push("/summary");
+  };
+
+  const Empty = () => {
+    return (
+      <>
+        <Header as="h2">{t("form.empty.title")}</Header>
+        <Trans i18nKey="form.empty.text">
```
Comment (mscharfstein) [inline client/src/components/Form.jsx:42]: what is a Trans component

File: client/src/components/renderWith.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            title: "form.title",
+          },
+        },
+      },
+    },
+  });
+
+const mockStore = configureStore();
+
+function renderWith(ui, { state, store = mockStore(state), ...options } = {}) {
```
Comment (kebrahim) [inline client/src/components/renderWith.js:36]: please add documentation

File: client/src/components/Form.jsx [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+  const form = useSelector((state) => state.form) || [];
+  const [page, setPage] = useState(0);
+  const size = form.length;
+
+  const isEmpty = size === 0;
+  const isFirstPage = page === 0;
+  const isFinalPage = page >= size;
+
+  const handleBack = () => {
+    // validate that for each questionKey in section, there is an equivalent
```
Comment (kebrahim) [inline client/src/components/Form.jsx:20]: is this a TODO to add this validation, or are you indicating that this validation occurs in the call to `setPage`?

File: client/src/components/Form.jsx [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+          )}
+          {isFinalPage && (
+            <Button onClick={handleDone} primary>
+              {t("form.done")}
+            </Button>
+          )}
+        </div>
+        <div>
+          {!isFirstPage && (
+            <Link to="" onClick={handleBack}>
```
Comment (mscharfstein) [inline client/src/components/Form.jsx:126]: Why link to ""? Should this be a Button component?

File: client/src/components/Form.jsx [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+  const isFirstPage = page === 0;
+  const isFinalPage = page >= size;
+
+  const handleBack = () => {
+    // validate that for each questionKey in section, there is an equivalent
+    // questionKey in responses; don't change page unless valid
+    setPage(Math.max(0, page - 1));
+  };
+
+  const handleNext = () => {
```
Comment (kebrahim) [inline client/src/components/Form.jsx:25]: how about `handleContinue` to match the button in the mocks?

File: client/src/components/Form.jsx [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    e.preventDefault();
+    // validate that for each questionKey in form, there is an equivalent
+    // questionKey in responses
+    history.push("/summary");
+  };
+
+  const Empty = () => {
+    return (
+      <>
+        <Header as="h2">{t("form.empty.title")}</Header>
```
Comment (kebrahim) [inline client/src/components/Form.jsx:41]: just curious - why h2 & not h1?

File: client/src/components/Form.jsx [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+  const handleDone = (e) => {
+    e.preventDefault();
+    // validate that for each questionKey in form, there is an equivalent
+    // questionKey in responses
+    history.push("/summary");
+  };
+
+  const Empty = () => {
+    return (
+      <>
```
Comment (kebrahim) [inline client/src/components/Form.jsx:40]: what is this magic? ;)

File: client/src/components/Form.jsx [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+    );
+  };
+
+  const Intro = () => {
+    return (
+      <>
+        <Header as="h2">{t("form.intro.title")}</Header>
+        <Trans i18nKey="form.intro.body">
+          <p>
+            We&apos;ll ask you a few, simple question about the your current
```
Comment (kebrahim) [inline client/src/components/Form.jsx:55]: s/question/questions

Comment (kebrahim) [inline client/src/components/Form.jsx:55]: isn't this string in the translation.json above? why can't you pull it from there?

File: client/src/components/Form.jsx [hunk 12]
```diff
@@ -0,0 +1,144 @@
+import React, { useState } from "react";
+import { Form as FormImpl, Container, Header, Button } from "semantic-ui-react";
+import { useTranslation, Trans } from "react-i18next";
+import { useSelector } from "react-redux";
+import { useHistory, Link } from "react-router-dom";
+import PropTypes from "prop-types";
+
+export default function Form() {
```
Comment (kebrahim) [inline client/src/components/Form.jsx:8]: should we also delete QuestionForm.jsx & QuestionForm.test.js?

File: client/src/components/Form.jsx [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+      <nav>
+        <div>
+          {isFirstPage && (
+            <Button onClick={handleNext} primary>
+              {t("form.init")}
+            </Button>
+          )}
+          {!isFirstPage && !isFinalPage && (
+            <Button onClick={handleNext} primary>
+              {t("form.next")}
```
Comment (kebrahim) [inline client/src/components/Form.jsx:115]: `form.continue` to match the mocks?

File: client/public/locales/en/translation.json [hunk 14]
```diff
@@ -1,10 +1,27 @@
...
+  "Please select yes or no": "Please select yes or no.",
+  "Select Language": "Select Language",
   "button": {
     "back": "Back",
     "submit": "Submit"
   },
   "form": {
-    "title": "Form"
+    "back": "Go back to previous step",
+    "done": "Done",
```
Comment (kebrahim) [inline client/public/locales/en/translation.json:12]: is this in the mocks? I don't see a "Done" button - https://preview.uxpin.com/642233d8d1d76a804e0ad41477dc05e51ccf21fc#/pages/129992668/specification/sitemap?mode=i

File: client/src/components/Form.test.js [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+      expect(back()).toBeInTheDocument();
+      expect(next()).toBeInTheDocument();
+      expect(next()).toHaveClass("primary");
+      expect(done()).not.toBeInTheDocument();
+    });
+
+    it("has back/done buttons on final page", async () => {
+      const { init, next, back, done } = queryButtons(screen);
+      fireEvent.click(init());
+      fireEvent.click(next());
```
Comment (kebrahim) [inline client/src/components/Form.test.js:116]: should we also test that clicking the back button in the case where we have both 1 & 2 sections works as expected?

File: client/src/components/Form.test.js [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+      expect(next()).not.toBeInTheDocument();
+      expect(back()).not.toBeInTheDocument();
+      expect(done()).not.toBeInTheDocument();
+    });
+  });
+
+  describe("with one section", () => {
+    beforeEach(() => {
+      const initialState = {
+        form: Array(1).fill({ section: [] }),
```
Comment (kebrahim) [inline client/src/components/Form.test.js:53]: this probably needs to wait until you're using Michelle's section component, but also confirm that the expected sectionKey & questionKeys are passed to the Section component

File: client/.eslintrc.js [hunk 17]
```diff
@@ -45,7 +45,13 @@ module.exports = {
       2,
       {
         markupOnly: true,
-        ignoreAttribute: ["path", "value"],
+        ignoreAttribute: ["path", "value", "as", "i18nKey"],
+      },
+    ],
+    "i18n-json/valid-message-syntax": [
```
[the remark this answers is not in the document]
Comment (marcja) [reply client/.eslintrc.js:51]: I turned off this rule. It can't handle i18next syntax with placeholders; it only supports ICU syntax.

File: client/public/locales/bn/translation.json [hunk 18]
```diff
@@ -1,9 +1,26 @@
 {
+  "Language dropdown": "Language dropdown",
```
[the remark this answers is not in the document]
Comment (marcja) [reply client/public/locales/bn/translation.json:2]: These are Sumer's strings.

[the remark this answers is not in the document]
Comment (marcja) [reply client/public/locales/bn/translation.json:2]: These are Sumer's strings. They should default to the translation key, not the English. But again, outside the scope of my PR.

File: client/public/locales/en/translation.json (hunk 14 above)
Comment (marcja) [reply client/public/locales/en/translation.json:12]: The final strings might have the same text for both next and done (that is, "Continue"). But the buttons are different and have different functionality, so I thought it made sense to support two different strings. It also made it easier to manually test.

File: client/public/locales/en/translation.json [hunk 19]
```diff
@@ -1,10 +1,27 @@
...
+    "empty": {
+      "text": "form.empty.text",
```
[the remark this answers is not in the document]
Comment (marcja) [reply client/public/locales/en/translation.json:14]: Yes. There is a linter rule that says all the translation files need to have the same keys. When you run `yarn run i18n:extract`, it synchronizes all the keys and defaults any newly created keys to the text of the key itself.

File: client/src/App.test.js (hunk 1 above)
Comment (marcja) [reply client/src/App.test.js:25]: No, I'll get these working before submit. But they were pretty busted at this time, so I turned them off.

File: client/src/Routes.test.js (hunk 2 above)
Comment (marcja) [reply client/src/Routes.test.js:14]: I'll check it again, but I think it's still busted.

File: client/src/components/Form.jsx (hunk 12 above)
Comment (marcja) [reply client/src/components/Form.jsx:8]: Will do before submit.

File: client/src/components/Form.jsx [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+export default function Form() {
+  const { t } = useTranslation();
+  const history = useHistory();
+  const form = useSelector((state) => state.form) || [];
+  const [page, setPage] = useState(0);
+  const size = form.length;
+
+  const isEmpty = size === 0;
+  const isFirstPage = page === 0;
+  const isFinalPage = page >= size;
```
[the remark this answers is not in the document]
Comment (marcja) [reply client/src/components/Form.jsx:17]: Nope. If the form data has four sections, the Form will display five pages (including the privacy policy intro page). This is correct.

File: client/src/components/Form.jsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (marcja) [reply client/src/components/Form.jsx:20]: Yep

Comment (marcja) [reply client/src/components/Form.jsx:20]: It's a TODO. Will add

File: client/src/components/Form.jsx (hunk 10 above)
Comment (marcja) [reply client/src/components/Form.jsx:40]: Nope. It is the [short syntax](https://reactjs.org/docs/fragments.html#short-syntax) for [React.Fragment](https://reactjs.org/docs/fragments.html).

File: client/src/components/Form.jsx (hunk 9 above)
Comment (marcja) [reply client/src/components/Form.jsx:41]: It's at the same level of visual hierarchy as section.title and intro.title.

File: client/src/components/Form.jsx (hunk 4 above)
Comment (marcja) [reply client/src/components/Form.jsx:42]: [Trans](https://react.i18next.com/latest/trans-component) is the component form of `t(...)`. The advantage here is that you can supply a few minimal HTML components (like `<p>`). `t(...)` is generally preferred except in cases like this.

File: client/src/components/Form.jsx (hunk 11 above)
Comment (marcja) [reply client/src/components/Form.jsx:55]: I can. This was a debugging convenience. Will remove.

File: client/src/components/Form.jsx (hunk 13 above)
Comment (marcja) [reply client/src/components/Form.jsx:115]: This is the translation key, not the text itself. form.text can be set to "Continue". Generally speaking, we don't need to change the translation key to match every time someone changes the text.

File: client/src/components/Form.jsx (hunk 7 above)
Comment (marcja) [reply client/src/components/Form.jsx:126]: Unfortunately, Semantic UI React doesn't support link-styled buttons yet. I tried. The mocks have this styled as a link, not a button.

File: client/src/components/Form.test.js (hunk 15 above)
Comment (marcja) [reply client/src/components/Form.test.js:116]: Good call. Haven't tested back yet.

File: client/src/components/renderWith.js (hunk 3 above)
Comment (marcja) [reply client/src/components/renderWith.js:1]: Yes, I plan to move it into a `utils` directory (possible under a `tests` directory that is sibling to `src`, but I had some issues including files outside of `src`).

File: client/src/components/renderWith.js (hunk 5 above)
Comment (marcja) [reply client/src/components/renderWith.js:36]: What, it's not self-explanatory? ;)

File: client/src/components/FormIntro.test.js [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+import FormIntro from "./FormIntro.jsx";
+import renderWith from "./renderWith";
+
+describe("form intro", () => {
+  beforeEach(() => {
+    renderWith(<FormIntro />);
+  });
+
+  it("has heading", async () => {
+    const heading = screen.getByRole("heading", { name: "form.intro.title" });
```
Comment (mscharfstein) [inline client/src/components/FormIntro.test.js:13]: does this work? trying the same thing and failing

**Comment by mscharfstein** [comment]
Updated code to use the real Section component and merge with validation changes. PTAL

File: client/src/components/Form.jsx (hunk 8 above)
Comment (marcja) [reply client/src/components/Form.jsx:36]: Prefer not. The "Continue" text in the mocks is variable.

File: client/src/components/Form.jsx [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+    }
+  };
+
+  const isValidAnswers = () => {
+    if (page === 0) {
+      // On FormEntry page, we do not need to check answers to any questions.
+      return true;
+    }
+    const sectionInfo = form[page - 1];
+    // eslint-disable-next-line
```
Comment (marcja) [inline client/src/components/Form.jsx:59]: You don't mean to disable all of eslint, just the unused vars rule, right? Change to:

```'js
// eslint-disable-next-line no-unused-vars
```

Comment (mscharfstein) [reply client/src/components/Form.jsx:59]: ack. i tried to make the change in the eslintrc itself and it seemed to work when running yarn lint but then when the tests ran on the PR it failed for that reason. updated the comment here

File: client/src/components/FormIntro.test.js [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+    expect(heading).toBeInTheDocument();
+    expect(heading).toHaveTextContent("form.intro.title");
+  });
+});
+
+test("form intro is accessible", async () => {
```
Comment (kebrahim) [inline client/src/components/FormIntro.test.js:22]: why isn't this part of the above describe/it block?

File: client/src/components/FormSteps.jsx [hunk 24]
```diff
@@ -0,0 +1,57 @@
+import PropTypes from "prop-types";
+import React from "react";
+import { useTranslation } from "react-i18next";
+import { Form } from "semantic-ui-react";
+import Section from "./Section.jsx";
+
+function FormSteps({ page, form, showError }) {
```
Comment (kebrahim) [inline client/src/components/FormSteps.jsx:7]: add documentation

File: client/src/components/FormSteps.jsx [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+            questionKeys={questionKeys}
+            showErrorForNoEntry={showError}
+          />
+        </Form.Group>
+      </Form>
+    </>
+  );
+}
+
+function nonNegativeInteger(props, propName, componentName) {
```
Comment (kebrahim) [inline client/src/components/FormSteps.jsx:32]: add documentation

File: client/src/components/FormSteps.test.js [hunk 26]
```diff
@@ -0,0 +1,21 @@
...
+    const initialState = {
+      form: Array(1).fill({ section: [] }),
+    };
+    renderWith(<FormSteps form={initialState.form} />, {
+      state: initialState,
+    });
+  });
+
+  it("has heading", async () => {
+    const div = screen.getByTestId("formStepMain");
```
Comment (kebrahim) [inline client/src/components/FormSteps.test.js:18]: also verify the correct `Step x of y` is displayed

File: client/src/components/LanguageSwitcher.jsx [hunk 27]
```diff
@@ -42,7 +42,6 @@ function LanguageSwitcher() {
         onChange={changeLanguage}
         placeholder={t("Select Language")}
         fluid
-        select
```
Comment (kebrahim) [inline client/src/components/LanguageSwitcher.jsx:45]: was this change intended?

Comment (marcja) [reply client/src/components/LanguageSwitcher.jsx:45]: yes -- it was an invalid attribute that caused compiler warning

File: client/src/components/FormSteps.test.js (hunk 26 above)
Comment (marcja) [reply client/src/components/FormSteps.test.js:18]: this is tricky

File: client/src/components/FormIntro.test.js (hunk 23 above)
Comment (mscharfstein) [reply client/src/components/FormIntro.test.js:22]: i didnt change the tests but i think maybe because it tests something separate (accessibility)?

Comment (marcja) [reply client/src/components/FormIntro.test.js:22]: Because these a11y/axe tests are the only type of tests that require the container passed back from render. This makes the pattern used elsewhere (`render` in `beforeAll` and access via `screen`) not applicable to a11y/axe tests. So I thought the cleanest way to do it was like this.

File: client/src/components/LanguageSwitcher.jsx (hunk 27 above)
Comment (mscharfstein) [reply client/src/components/LanguageSwitcher.jsx:45]: marc said yes he meant to otherwise build breaks

File: client/src/components/FormSteps.test.js [hunk 28]
```diff
@@ -9,7 +9,7 @@ describe("form steps", () => {
     const initialState = {
       form: Array(1).fill({ section: [] }),
     };
-    renderWith(<FormSteps form={initialState.form} />, {
+    renderWith(<FormSteps page={1} form={initialState.form} size={3} />, {
```
Comment (marcja) [inline client/src/components/FormSteps.test.js:12]: `size` isn't a valid prop

File: client/src/components/FormSteps.test.js (hunk 26 above)
Comment (mscharfstein) [reply client/src/components/FormSteps.test.js:18]: yeah, it looks like i can verify that form.page is displayed but not the actual translation of that using i18next?

File: client/src/components/FormSteps.test.js (hunk 28 above)
Comment (mscharfstein) [reply client/src/components/FormSteps.test.js:12]: ack

File: client/src/components/FormSteps.jsx [hunk 29]
```diff
@@ -29,6 +38,9 @@ function FormSteps({ page, form, showError }) {
   );
 }
 
+/**
+ * Checks that the current page index is a non-negative number.
```
Comment (kebrahim) [inline client/src/components/FormSteps.jsx:42]: it's not always current page index, right? it's whatever is passed in as propName

File: client/src/components/FormSteps.test.js (hunk 26 above)
Comment (marcja) [reply client/src/components/FormSteps.test.js:18]: The way I was going to test this was by creating on "onPageChange" prop that can be mocked. So:

```js
const handlePageChange = jest.fn();
renderWith(<FormSteps onPageChange={handlePageChange} />)
fireEvent.click(...);
expect(handlePageChange).toHaveBeenCalledWith(1);
fireEvent.click(...);
expect(handlePageChange).toHaveBeenCalledWith(2);
```

File: client/src/components/Form.jsx [hunk 30]
```diff
@@ -0,0 +1,112 @@
...
+    const [_, questionKeys] = Object.entries(sectionInfo)[0];
```
[the remark this answers is not in the document]
Comment (marcja) [reply client/src/components/Form.jsx:66]: Kurt FTW

File: client/src/components/FormIntro.jsx [hunk 31]
```diff
@@ -0,0 +1,29 @@
...
+
+const FormIntro = () => {
+  const { t } = useTranslation();
+
+  return (
+    <div>
+      <Header as="h2">{t("form.intro.title")}</Header>
+      <Trans i18nKey="form.intro.body">
+        <p>
+          We&apos;ll ask you a few, simple question about the your current
```
[the remark this answers is not in the document]
Comment (marcja) [reply client/src/components/FormIntro.jsx:13]: The string is translation.json was already created. This text can be removed. But we need to keep using the <Trans> component or else the <p> tags will be escaped and displayed as "<p>" (unless there's another way to do that...)

File: client/src/components/FormSteps.jsx [hunk 32]
```diff
@@ -0,0 +1,69 @@
...
+  const size = form.length;
+  const formPage = form[page - 1];
+  const [sectionKey, questionKeys] = Object.entries(formPage)[0];
+
+  return (
+    <>
+      <div data-testid="formStepMain">
+        {t("form.page", { page: page, size: size })}
+      </div>
+      <Form>
```
Comment (marcja) [inline client/src/components/FormSteps.jsx:28]: Should the showError prop be added to Form? For example:

```js
<Form error={showError}>
```

According to https://react.semantic-ui.com/collections/form/, this means that Form will "Automatically show any error Message children.".

File: client/src/components/Form.jsx (hunk 30 above)
[the remark this answers is not in the document]
Comment (mscharfstein) [reply client/src/components/Form.jsx:66]: oops sorry i missed this one. didnt realize i could do that. updated!

File: client/src/components/FormIntro.jsx (hunk 31 above)
[the remark this answers is not in the document]
Comment (mscharfstein) [reply client/src/components/FormIntro.jsx:13]: done

File: client/src/components/FormSteps.jsx [hunk 33]
```diff
@@ -4,6 +4,15 @@ import { useTranslation } from "react-i18next";
 import { Form } from "semantic-ui-react";
 import Section from "./Section.jsx";
 
+/**
+ * @param {object} props
+ *   - page: an index indicating which page of questions the user is on
+ *   - form: an array of objects that holds section information
+ *   - showError: a boolean value indicating whether to display
+ *     an error prompt if a question has not been answered. Defaults
+ *     to false. Will be true once the user attempts to submit the form.
```
[the remark this answers is not in the document]
Comment (mscharfstein) [reply client/src/components/FormSteps.jsx:13]: showError will be true if at least one question is not answered. updated

File: client/src/components/FormSteps.test.js (hunk 26 above)
Comment (mscharfstein) [reply client/src/components/FormSteps.test.js:18]: ok, will leave for another PR
